src/Servers/Kestrel/Core/src/Internal/Http2/FlowControl/InputFlowControl.cs

```diff
@@ -19,94 +19,125 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.FlowControl;
     private int _pendingUpdateSize;
     private bool _windowUpdatesDisabled;
-    private readonly object _flowLock = new object();
 
     public InputFlowControl(uint initialWindowSize, uint minWindowSizeIncrement)
     {
         Debug.Assert(initialWindowSize >= minWindowSizeIncrement, "minWindowSizeIncrement is greater than the window size.");
 
-        _flow = new FlowControl(initialWindowSize);
-        _initialWindowSize = (int)initialWindowSize;
+        _flow = new FlowControlState(initialWindowSize, false);
+        _initialWindowSize = initialWindowSize;
         _minWindowSizeIncrement = (int)minWindowSizeIncrement;
     }
 
     public bool IsAvailabilityLow => _flow.Available < _minWindowSizeIncrement;
 
+    // Test hook, not participating in mutual exclusion
+    internal uint Available => _flow.Available;
+
     public void Reset()
     {
-        _flow = new FlowControl((uint)_initialWindowSize);
+        _flow = new FlowControlState(_initialWindowSize, false);
         _pendingUpdateSize = 0;
         _windowUpdatesDisabled = false;
     }
 
     public bool TryAdvance(int bytes)
     {
-        lock (_flowLock)
+        FlowControlState currentFlow, computedFlow;
+        do
         {
+            currentFlow = _flow; // Copy
             // Even if the stream is aborted, the client should never send more data than was available in the
             // flow-control window at the time of the abort.
-            if (bytes > _flow.Available)
+            if (bytes > currentFlow.Available)
             {
                 throw new Http2ConnectionErrorException(CoreStrings.Http2ErrorFlowControlWindowExceeded, Http2ErrorCode.FLOW_CONTROL_ERROR, ConnectionEndReason.FlowControlWindowExceeded);
             }
 
-            if (_flow.IsAborted)
+            if (currentFlow.IsAborted)
             {
-                // This data won't be read by the app, so tell the caller to count the data as already consumed.
                 return false;
             }
 
-            _flow.Advance(bytes);
-            return true;
-        }
+            computedFlow = new FlowControlState(currentFlow.Available - (uint)bytes, isAborted: false);
+        } while (currentFlow._state != Interlocked.CompareExchange(ref _flow._state, computedFlow._state, currentFlow._state));
+
+        return true;
     }
 
     public bool TryUpdateWindow(int bytes, out int updateSize)
     {
-        lock (_flowLock)
+        FlowControlState currentFlow, computedFlow;
+        do
         {
             updateSize = 0;
-
-            if (_flow.IsAborted)
+            currentFlow = _flow; // Copy
+            if (currentFlow.IsAborted)
             {
                 // All data received by stream has already been returned to the connection window.
                 return false;
             }
 
-            if (!_flow.TryUpdateWindow(bytes))
+            var maxUpdate = int.MaxValue - currentFlow.Available;
+            if (bytes > maxUpdate)
```

Comment:
Can this condition go (reversed) in the assert?

Reply:
yes, also updating `int.MaxValue` the line above, which is wrong.